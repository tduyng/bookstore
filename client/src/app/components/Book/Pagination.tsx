import React, { useEffect, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';

interface PaginationProps {
  count: number;
  limit: number;
  page: number;
  q?: string;
}

export const Pagination: React.FC<PaginationProps> = ({ count, limit, page, q }) => {
  const [currentPage, setCurrentPage] = useState(page);
  const pages = Math.ceil(count / limit);
  const location = useLocation();
  const path = location.pathname;

  const range = (start: number, stop: number, step: number) =>
    Array.from({ length: (stop - start) / step + 1 }, (_, i) => start + i * step);

  const changePage = (e: any) => {
    window.scroll({ top: 0, behavior: 'smooth' });
    setCurrentPage(parseInt(e.target.textContent));
  };

  const nextPage = () => {
    window.scroll({ top: 0, behavior: 'smooth' });
    setCurrentPage(currentPage + 3);
  };

  const prevPage = () => {
    window.scroll({ top: 0, behavior: 'smooth' });
    setCurrentPage(currentPage - 3);
  };

  useEffect(() => {
    setCurrentPage(page);
  }, [path, page]);

  return (
    <div className="pagination">
      <Link
        className={`pagination__page${currentPage === 1 ? ' active' : ''}`}
        onClick={changePage}
        key={1}
        to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=1&limit=${limit}`}
      >
        1
      </Link>
      {/* Icon prev */}
      {currentPage > 4 && pages > 5 && (
        <Link
          className="pagination__prev-page"
          onClick={prevPage}
          to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=${
            currentPage - 3
          }&limit=${limit}`}
        >
          <i className="fas fa-angle-double-left"></i>
        </Link>
      )}
      {range(
        currentPage >= 4 ? currentPage - 2 : 2,
        currentPage + 2 >= pages - 1 ? pages - 1 : currentPage + 2,
        1,
      ).map(n => (
        <Link
          className={`pagination__page${currentPage === n ? ' active' : ''}`}
          onClick={changePage}
          key={n}
          to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=${n}&limit=${limit}`}
        >
          {n}
        </Link>
      ))}

      {/* Icon next */}
      {currentPage <= pages - 4 && pages > 5 && (
        <Link
          className="pagination__next-page"
          onClick={nextPage}
          to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=${
            currentPage + 3
          }&limit=${limit}`}
        >
          <i className="fas fa-angle-double-right"></i>
        </Link>
      )}
      <Link
        className={`pagination__page${currentPage === pages ? ' active' : ''}`}
        onClick={changePage}
        key={pages}
        to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${
          q ? '&' : ''
        }p=${pages}&limit=${limit}`}
      >
        {pages}
      </Link>
    </div>
  );
};
