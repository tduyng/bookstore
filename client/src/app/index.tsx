import * as React from 'react';
import { Helmet } from 'react-helmet-async';
import { Toaster } from 'react-hot-toast';
import { Switch, Route, BrowserRouter } from 'react-router-dom';
import { PATH } from './constants/paths.constant';
import { HomePage } from './pages/HomePage/Loadable';

import { NotFoundPage } from './pages/NotFoundPage/Loadable';

export function App() {
  return (
    <BrowserRouter>
      <Helmet titleTemplate="%s - Bookstore" defaultTitle="Bookstore">
        <meta name="description" content="Bookstore application" />
      </Helmet>
      <Toaster
        position="top-right"
        reverseOrder={false}
        toastOptions={{
          style: { fontSize: '1.6rem' },
        }}
      />

      <Switch>
        {/* <Route path="/books/genre/:genre" component={BooksGenre} />
        <Route path="/books/search" component={BooksSearch} />
        <Route path="/account" component={Account} />
        <Route path="/book/:id" component={BookDetail} />
        <Route path="/checkout/payment" component={Payment} />
        <Route path="/checkout" component={Checkout} exact /> */}
        <Route exact path={PATH.HOME} component={HomePage} />
        <Route component={NotFoundPage} />
      </Switch>
    </BrowserRouter>
  );
}

export default App;
