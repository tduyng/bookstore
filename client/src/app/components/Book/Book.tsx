import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { IBook } from 'src/app/features/book/book.types';
import { addToCart } from 'src/app/features/user/user.actions';
import { CartItemDto } from 'src/app/features/user/user.types';
import { useAppDispatch } from 'src/store';
import toast from 'react-hot-toast';

type BookProps = IBook & { hideButton?: boolean };
export const Book: React.FC<BookProps> = props => {
  const dispatch = useAppDispatch();

  const { _id, title, price, old_price, imgURL } = props;
  const [btnText, setBtnText] = useState('Add to cart');

  const handleAddToCart = ({ _id, total }: CartItemDto) => {
    setBtnText('Adding ...');
    dispatch(addToCart({ _id, total })).then(unwrapResult => {
      setBtnText('Add to cart');
      if (unwrapResult.meta.requestStatus === 'fulfilled') {
        toast.success('Added product to Cart');
      } else {
        toast.error(`Can't add product to cart, please try later.`);
      }
    });
  };

  return (
    <div className="book">
      <Link to={`/book/${_id}`} className="book__link">
        <img className="book__link--img" src={imgURL} alt=""></img>
      </Link>
      <div className="book__content">
        <div className="book__content--name">{title}</div>
        <div className="book__content--price">
          {(price * 1000).toLocaleString('vi-VN', {
            style: 'currency',
            currency: 'VND',
          })}
        </div>
        {old_price && (
          <div className="book__content--old-price">
            {(old_price * 1000).toLocaleString('vi-VN', {
              style: 'currency',
              currency: 'VND',
            })}
          </div>
        )}

        {props.hideButton ? null : (
          <button
            onClick={() => handleAddToCart({ _id, total: 1 })}
            className="book__content--button"
          >
            {btnText}
          </button>
        )}
      </div>
    </div>
  );
};
