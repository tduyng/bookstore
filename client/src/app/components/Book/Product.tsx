import React, { useState } from 'react';
import toast from 'react-hot-toast';
import { Link } from 'react-router-dom';
import { removeFromCart } from 'src/app/features/user/user.actions';
import { CartItem } from 'src/app/features/user/user.types';
import { useAppDispatch } from 'src/store';
import noImage from '/no-image.png';

export const Product: React.FC<CartItem> = props => {
  const dispatch = useAppDispatch();
  const id = props._id as string;
  const [inpValue, setInpValue] = useState(props.total || 1);
  const [amount, setAmount] = useState(props.total || 1);

  const formatter = new Intl.NumberFormat('vi-VN', {
    style: 'currency',
    currency: 'VND',
  });

  const handleChange = (e: any) => {
    if (isNaN(e.target.value)) {
      setInpValue(amount);
    } else if (parseInt(e.target.value) > 99) {
      setInpValue(99);
    } else {
      setInpValue(e.target.value);
    }
  };

  const updateAmount = (e: any) => {
    if (
      !e.target.value ||
      parseInt(e.target.value) === 0 ||
      parseInt(e.target.value) > 99
    ) {
      setInpValue(amount);
    } else {
      setAmount(parseInt(e.target.value));
      setInpValue(parseInt(e.target.value));
    }
  };

  const deleteProduct = (_id: string) => {
    dispatch(removeFromCart({ _id })).then(result => {
      if (result.meta.requestStatus === 'fulfilled') {
        toast.success('Remove cart item');
      }
    });
  };

  return (
    <div className="checkout__product">
      <i className="fas fa-times" onClick={() => deleteProduct(id)}></i>
      <Link to={`/book/${id}`} className="checkout__product--img">
        <img src={props.imgURL || noImage} alt="Product" />
      </Link>
      <div className="checkout__product--detail">
        <p className="checkout__product--detail__name">{props.title}</p>
        <p className="checkout__product--detail__price">
          {formatter.format(
            parseFloat(props.price ? props.price.toString() : '0') * 1000,
          )}
        </p>
        {props.old_price && (
          <p className="checkout__product--detail__old-price">
            {formatter.format(parseFloat(props.old_price.toString()) * 1000)}
          </p>
        )}
      </div>
      <div className="checkout__product--amount">
        <input
          type="text"
          value={inpValue}
          onChange={handleChange}
          onBlur={updateAmount}
        />
        <p className="checkout__product--amount__total">Total:</p>
        <p className="checkout__product--amount__money">
          {formatter.format(inpValue * (props.price ? props.price * 1000 : 0))}
        </p>
      </div>
    </div>
  );
};

Product.defaultProps = {
  total: 1,
};
