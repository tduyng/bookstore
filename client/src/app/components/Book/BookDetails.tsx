import React, { useEffect, useState } from 'react';
import toast from 'react-hot-toast';
import { useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router';
import Slider from 'react-slick';
import { fetchBookById, fetchBooks } from 'src/app/features/book/book.actions';
import { IBook } from 'src/app/features/book/book.types';
import { addToCart } from 'src/app/features/user/user.actions';
import { NotFoundPage } from 'src/app/pages/NotFoundPage/Loadable';
import { useAppDispatch } from 'src/store';
import { AppState } from 'src/store/reducers';
import SkeletonBook from '../skeleton/SkeletonBook';
import { Book } from './Book';
import noImage from '/no-image.png';
interface ArrowProps {
  onClick?: () => void;
}
const NextArrow: React.FC<ArrowProps> = props => {
  const { onClick } = props;
  return (
    <div className={'icon-related next'} onClick={onClick}>
      <i className="fas fa-arrow-circle-right"></i>
    </div>
  );
};

const PrevArrow: React.FC<ArrowProps> = props => {
  const { onClick } = props;
  return (
    <div className={'icon-related prev'} onClick={onClick}>
      <i className="fas fa-arrow-circle-left"></i>
    </div>
  );
};

export const BookDetail = () => {
  const settings = {
    dots: false,
    infinite: true,
    slidesToShow: 6,
    slidesToScroll: 6,
    nextArrow: <NextArrow />,
    prevArrow: <PrevArrow />,
    responsive: [
      {
        breakpoint: 1000,
        settings: {
          slidesToShow: 5,
          slidesToScroll: 5,
        },
      },
      {
        breakpoint: 768,
        settings: {
          slidesToShow: 4,
          slidesToScroll: 4,
        },
      },
      {
        breakpoint: 510,
        settings: {
          slidesToShow: 3,
          slidesToScroll: 3,
        },
      },
      {
        breakpoint: 430,
        settings: {
          slidesToShow: 2,
          slidesToScroll: 2,
        },
      },
    ],
  };

  const { id } = useParams<{ id: string }>();
  const dispatch = useAppDispatch();
  const { book, status, message, books } = useSelector((state: AppState) => state.book);
  const [amount, setAmount] = useState(1);
  const [btnText, setBtnText] = useState('Add to Cart');
  const [foundBook, setFoundBook] = useState(false);
  const history = useHistory();

  const handleAddToCart = (id: string, amount: number) => {
    setBtnText('Adding...');
    dispatch(addToCart({ _id: id, total: amount })).then(result => {
      setBtnText('Add to cart');
      if (result.meta.requestStatus === 'fulfilled') {
        toast.success('Added product to cart');
      } else {
        toast.error(`Can't add product to cart, please try later`);
      }
    });
  };
  const add = () => {
    setAmount(amount + 1);
  };
  const minus = () => {
    if (amount > 1) {
      setAmount(amount - 1);
    }
  };

  useEffect(() => {
    dispatch(fetchBookById(id)).then(result => {
      if (result.meta.requestStatus === 'fulfilled') {
        setFoundBook(true);
        const bookFound = result.payload as IBook;
        const genre = (bookFound ? bookFound.genre : '') as string;
        const limit = 25;
        const p = 1;
        dispatch(fetchBooks({ genre, limit, page: p }));
      } else {
        setFoundBook(false);
      }
    });
  }, [id, dispatch, foundBook, setFoundBook]);

  if (!foundBook) {
    return (
      <div className="book-detail__wrapper">
        <div className="book-detail">
          <div className="not-found">
            <h1>Sorry</h1>
            <p>That page can't be found</p>
            <a className="not-found__link" onClick={() => history.goBack()}>
              Go back
            </a>
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="book-detail__wrapper">
      <div className="book-detail">
        {(status === 'idle' || status === 'loading') && <SkeletonBook />}
        {status === 'success' && (
          <>
            <div className="book-detail__media">
              <img
                src={book?.imgURL || noImage}
                alt="Book"
                className="book-detail__media--img"
              />
            </div>
            <div className="book-detail__content">
              <h1 className="book-detail__content--title">{book?.title}</h1>

              <p className="book-detail__content--author">
                Author: <span>{book?.author}</span>
              </p>

              <div className="book-detail__content--price">
                <p className="book-detail__content--price__special">
                  {(book?.price ? book.price * 1000 : 0).toLocaleString('vi-VN', {
                    style: 'currency',
                    currency: 'VND',
                  })}
                </p>
                {book?.old_price && (
                  <p className="book-detail__content--price__old">
                    {(book?.old_price * 1000).toLocaleString('vi-VN', {
                      style: 'currency',
                      currency: 'VND',
                    })}
                  </p>
                )}
              </div>

              <div className="book-detail__content--amount">
                <p>Amount:</p>
                <div className="book-detail__content--amount__btn">
                  <i className="fas fa-minus" onClick={minus}></i>
                  <span>{amount}</span>
                  <i className="fas fa-plus" onClick={add}></i>
                </div>
              </div>

              <button
                className="book-detail__content--add"
                onClick={() => handleAddToCart(id, amount)}
              >
                {btnText}
              </button>
            </div>
          </>
        )}
        {status === 'failed' && <h1 className="book-detail__error">{message}</h1>}
      </div>

      {status === 'success' && (
        <div className="related">
          <p className="related__title">Related</p>
          <div className="related__slider--wrapper">
            <Slider {...settings} className="related__slider">
              {books.length
                ? books.map(item => (
                    <Book
                      imgURL={item.imgURL}
                      title={item.title}
                      price={item.price}
                      old_price={item.old_price}
                      key={item._id}
                      _id={item._id}
                      hideButton={true}
                    />
                  ))
                : null}
            </Slider>
          </div>
        </div>
      )}
    </div>
  );
};
