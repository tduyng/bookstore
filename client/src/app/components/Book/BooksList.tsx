import React, { Suspense, useEffect, useRef, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { IBook } from 'src/app/features/book/book.types';
import LazyLoad from 'react-lazyload';
import SkeletonBook from '../Skeletons/SkeletonBook';
import { Book } from './Book';
import { Pagination } from './Pagination';

interface BooksListProps {
  status: string;
  books: IBook[];
  page: number;
  count: number;
  message: string;
  limit: number;
  q?: string;
}
export const BooksList: React.FC<BooksListProps> = props => {
  const { status, books, page, count, message, limit, q } = props;
  const [activeLimit, setActiveLimit] = useState(false);
  const [path, setPath] = useState('');
  const location = useLocation();
  const handleClickLimit = () => {
    setActiveLimit(!activeLimit);
  };
  const ref: any = useRef();
  const handleClickOutsideLimit = (e: any) => {
    if (ref.current?.classList?.contains('active') && !ref.current?.contains(e.target)) {
      setActiveLimit(!activeLimit);
    }
  };

  useEffect(() => {
    setPath(location.pathname);
    document.addEventListener('click', handleClickOutsideLimit);
    return () => {
      document.removeEventListener('click', handleClickOutsideLimit);
    };
  }, [path, setPath]);

  return (
    <>
      {status === 'success' && (
        <div className="books__sort">
          <div
            className={`books__sort--limit${activeLimit ? ' active' : ''}`}
            onClick={handleClickLimit}
            ref={ref}
          >
            <p className="books__sort--limit__current">{`${limit} products`}</p>
            <ul className="books__sort--limit__option" onMouseLeave={handleClickLimit}>
              <Link
                to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=1&limit=15`}
                className="books__sort--limit__option--link"
              >
                <li>15 products</li>
              </Link>
              <Link
                to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=1&limit=25`}
                className="books__sort--limit__option--link"
              >
                <li>25 products</li>
              </Link>
              <Link
                to={`${path}?${q ? 'q=' : ''}${q ? q : ''}${q ? '&' : ''}p=1&limit=50`}
                className="books__sort--limit__option--link"
              >
                <li>50 products</li>
              </Link>
            </ul>

            <div className="books__sort--limit__icon">
              <i className="fas fa-angle-down"></i>
            </div>
          </div>
        </div>
      )}
      {status === 'success' &&
        books.map(book => (
          <LazyLoad key={book._id} placeholder={<SkeletonBook />}>
            <Suspense fallback={<SkeletonBook />}>
              <Book
                imgURL={book.imgURL}
                title={book.title}
                price={book.price}
                old_price={book.old_price}
                key={book._id}
                _id={book._id}
              />
            </Suspense>
          </LazyLoad>
        ))}
      {status === 'failed' && <h1 className="error">{message}</h1>}
      {count > limit && <Pagination count={count} limit={limit} page={page} q={q} />}
    </>
  );
};
