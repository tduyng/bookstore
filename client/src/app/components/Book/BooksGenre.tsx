import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { fetchBooks } from 'src/app/features/book/book.actions';
import { useQueryUrl } from 'src/app/hooks/useQueryUrl';
import { useAppDispatch } from 'src/store';
import { AppState } from 'src/store/reducers';
import SkeletonBook from '../skeleton/SkeletonBook';
import { BooksList } from './BooksList';

export const BooksGenre = () => {
  const dispatch = useAppDispatch();
  const { books, status, count, message } = useSelector((state: AppState) => state.book);
  const { genre } = useParams<{ genre: string }>();
  const limit: number = parseInt(useQueryUrl('limit')) || 25;
  const page: number = parseInt(useQueryUrl('p')) || 1;

  useEffect(() => {
    const getBooks = async () => {
      await dispatch(fetchBooks({ genre, limit, page }));
    };
    getBooks();
  }, [genre, limit, page, dispatch]);

  return (
    <div className="books">
      {status === 'idle' || status === 'loading' ? (
        Array.from(Array(10).keys()).map(n => <SkeletonBook key={n} />)
      ) : (
        <>
          {status === 'success' && <div className="books__desc">{books[0].genre}</div>}
          <BooksList
            status={status}
            books={books}
            message={message}
            count={count}
            page={page}
            limit={limit}
          />
        </>
      )}
    </div>
  );
};
