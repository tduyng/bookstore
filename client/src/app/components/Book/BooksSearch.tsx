import React, { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { queryBooks } from 'src/app/features/book/book.actions';
import { useQueryUrl } from 'src/app/hooks/useQueryUrl';
import { useAppDispatch } from 'src/store';
import { AppState } from 'src/store/reducers';
import SkeletonBook from '../Skeletons/SkeletonBook';
import { BooksList } from './BooksList';

export const BookSearch = () => {
  const dispatch = useAppDispatch();
  const { status, count, books, message } = useSelector((state: AppState) => state.book);

  const limit = parseInt(useQueryUrl('limit')) || 25;
  const page = parseInt(useQueryUrl('p')) || 1;
  const q = useQueryUrl('q') || '';

  useEffect(() => {
    if (!q) return;
    const getQueriedBooks = async () => {
      return await dispatch(queryBooks({ q, limit, page }));
    };
    getQueriedBooks();
  }, [limit, page, dispatch, q]);
  return (
    <div className="books">
      {status === 'idle' || status === 'loading' ? (
        Array.from(Array(10).keys()).map(n => <SkeletonBook key={n} />)
      ) : (
        <>
          {status === 'success' && <div className="books__desc">Result for : {q}</div>}
          {books && books.length > 0 ? (
            <BooksList
              status={status}
              books={books}
              message={message}
              count={count}
              page={page}
              limit={limit}
              q={q}
            />
          ) : (
            <h2>No result found</h2>
          )}
        </>
      )}
    </div>
  );
};
