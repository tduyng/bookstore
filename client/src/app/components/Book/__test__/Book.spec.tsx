import React from 'react';
import { Provider } from 'react-redux';
import { store } from 'src/store';
import { Book, BookProps } from '../Book';
import { MemoryRouter } from 'react-router';
import renderer from 'react-test-renderer';

const renderComponent = (props: BookProps) =>
  renderer.create(
    <MemoryRouter>
      <Provider store={store}>
        <Book {...props} />
      </Provider>
    </MemoryRouter>,
  );

const mockBookProps: BookProps = {
  _id: 'some',
  title: 'some',
  price: 0,
  old_price: 0,
};
describe('<Book>', () => {
  it('Should match snapshot', () => {
    const bookComponent = renderComponent(mockBookProps);
    expect(bookComponent.toJSON()).toMatchSnapshot();
  });
});
