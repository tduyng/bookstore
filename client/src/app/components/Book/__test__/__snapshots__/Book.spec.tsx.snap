// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Book> Should match snapshot 1`] = `
<div
  className="book"
>
  <a
    className="book__link"
    href="/book/some"
    onClick={[Function]}
  >
    <img
      alt=""
      className="book__link--img"
      src="some-stub"
    />
  </a>
  <div
    className="book__content"
  >
    <div
      className="book__content--name"
    >
      some
    </div>
    <div
      className="book__content--price"
    >
      0 ₫
    </div>
    0
    <button
      className="book__content--button"
      onClick={[Function]}
    >
      Add to cart
    </button>
  </div>
</div>
`;
